### PR TITLE
Document Calico CNI tweaks

### PR DESCRIPTION
**Pull request by sridhargaddam**
Fixes issue: https://github.com/submariner-io/submariner-website/issues/257

Signed-Off-by: Sridhar Gaddam <sgaddam@redhat.com>

### CODE REVIEW

**Review by nyechiel — COMMENTED** [review]
This detailed technical explanation is great, but maybe it's too much to include in the website? I'd love to get feedback from others, but perhaps it's better to just include the solution ("For Calico support, please
instruct Calico, via IPPools, not to perform NAT on subnets associated with remote Pods and Services").

**Comment by sridhargaddam** [comment]
> This detailed technical explanation is great, but maybe it's too much to include in the website? I'd love to get feedback from others, but perhaps it's better to just include the solution ("For Calico support, please
> instruct Calico, via IPPools, not to perform NAT on subnets associated with remote Pods and Services").

Ack, I'm fine with it. I can move the information I included to the corresponding issue (for anyone interested to know the detailed steps).

CC: @mangelajo @skitt @tpantelis

**Comment by dfarrell07** [comment]
I don't mind having the technical details, but imho it shouldn't be a full sidebar item. Maybe it would make sense as a section under "Deployment", as it is a facet of deployments.

File: src/content/cnipluginmatrix/_index.en.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
++++
+title = "CNI Plugin Matrix"
```
Comment (mangelajo) [inline src/content/cnipluginmatrix/_index.en.md:2]:
```suggestion
title = "Network Plugin Matrix"
```

CNI is the "API" as far as I understood, but we tend to interchange both terms.

File: src/content/cnipluginmatrix/_index.en.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+date = 2020-02-19T20:03:44+01:00
+weight = 20
+pre = "<b>4. </b>"
++++
+
+{{< include "cnipluginmatrix/support_matrix.md" >}}
```
Comment (mangelajo) [inline src/content/cnipluginmatrix/_index.en.md:8]:
```suggestion
{{< include "networkpluginmatrix/support_matrix.md" >}}
```

File: src/content/cnipluginmatrix/support_matrix.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+* AWS
+* VMware vSphere
+* OpenStack
+* GCP
+
+### Kubernetes CNI Plugins
+
+Presently, Submariner has been tested with the following CNI Plugins that leverage kube-proxy with iptables mode:
+
+* [openshift-sdn](https://github.com/openshift/sdn)
```
Comment (mangelajo) [inline src/content/cnipluginmatrix/support_matrix.md:19]:
```suggestion
* [OpenShift-SDN](https://github.com/openshift/sdn)
```

I know it's just being moved, but a good opportunity to update it.

File: src/content/quickstart/_index.en.md [hunk 4]
```diff
@@ -51,31 +51,7 @@ An example of three clusters configured to use with Submariner (without GlobalNe
...
-
-### Kubernetes CNI Plugins
-
-Presently, Submariner has been tested with the following CNI Plugins that leverage kube-proxy with iptables mode:
-
-* [openshift-sdn](https://github.com/openshift/sdn)
-* [Weave](https://github.com/weaveworks/weave)
-* [Flannel](https://github.com/coreos/flannel)
-* [Canal](https://docs.projectcalico.org/getting-started/kubernetes/flannel/flannel)
+{{< include "cnipluginmatrix/support_matrix.md" >}}
```
Comment (mangelajo) [inline src/content/quickstart/_index.en.md:54]: May be we can add a link to the section instead? so we don't duplicate content which can be a penalty to crawlers...

**Review by mangelajo — CHANGES_REQUESTED** [review]
some comments.

**Comment by sridhargaddam** [comment]
> I don't mind having the technical details, but imho it shouldn't be a full sidebar item. Maybe it would make sense as a section under "Deployment", as it is a facet of deployments.

@mangelajo thanks for the review comments. What is your opinion on the following ^^?

**Comment by mangelajo** [comment]
> > I don't mind having the technical details, but imho it shouldn't be a full sidebar item. Maybe it would make sense as a section under "Deployment", as it is a facet of deployments.
> 
> @mangelajo thanks for the review comments. What is your opinion on the following ^^?

Sorry, it seems like my comment got lost in github.

I said that it was actually my idea to make it more prominent and obvious what we support, and how do we support it. @dfarrell07  but I'm good either way. @sridhargaddam @dfarrell07 what do you think?

**Comment by sridhargaddam** [comment]
> I said that it was actually my idea to make it more prominent and obvious what we support, and how do we support it. @dfarrell07 but I'm good either way. @sridhargaddam @dfarrell07 what do you think?

I don't have any strong preference about the location as both have their own merits IMO. 
As @dfarrell07 said, its part of deployment, so shall we move it to "Deployment" section? @nyechiel your thoughts on this please.

**Comment by nyechiel** [comment]
> > I said that it was actually my idea to make it more prominent and obvious what we support, and how do we support it. @dfarrell07 but I'm good either way. @sridhargaddam @dfarrell07 what do you think?
> 
> I don't have any strong preference about the location as both have their own merits IMO.
> As @dfarrell07 said, its part of deployment, so shall we move it to "Deployment" section? @nyechiel your thoughts on this please.

Deployment sounds reasonable to me. Let's get it merged :+1:

File: src/content/cnipluginmatrix/support_matrix.md (hunk 3 above)
Comment (sridhargaddam) [reply src/content/cnipluginmatrix/support_matrix.md]: Done

File: src/content/quickstart/_index.en.md [hunk 5]
```diff
@@ -72,10 +72,11 @@ this document.
 
 Presently, Submariner has been tested with the following CNI Plugins that leverage kube-proxy with iptables mode:
 
-* [openshift-sdn](https://github.com/openshift/sdn)
+* [Openshift-SDN](https://github.com/openshift/sdn)
 * [Weave](https://github.com/weaveworks/weave)
 * [Flannel](https://github.com/coreos/flannel)
 * [Canal](https://docs.projectcalico.org/getting-started/kubernetes/flannel/flannel)
+* [Calico](https://www.projectcalico.org/). Please refer to the [following](../deployment/calico/) section for deployment instructions.
```
Comment (nyechiel) [inline src/content/quickstart/_index.en.md:79]:
```suggestion
* [Calico](https://www.projectcalico.org/). Please refer to the [following section](../deployment/calico/) for deployment instructions.
```

File: src/content/deployment/calico/_index.en.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+weight = 30
++++
+
+Normally, the local CNI in your K8s cluster (based on Kube-proxy with iptables mode) programs iptable
+rules for networking of the Pods. When a Pod in your cluster tries to talk to an external-IP, the CNI
+generally performs MASQ/SNAT of the traffic as it does not belong to the local cluster.
+Similar to CNI, Submariner also programs certain IPTable rules and it requires these rules to be applied
+prior to the CNI rules in order to avoid the SNAT for inter-cluster traffic.
+Submariner tries to preserve the source-IP of the Pods for cross-cluster communication for visibility,
+ease of debugging, and for network policy support.
```
Comment (nyechiel) [inline src/content/deployment/calico/_index.en.md:12]:
```suggestion
Typically, the Kubernetes network plugin (based on kube-proxy) programs iptables
rules for Pods networking within the cluster. When a Pod in the cluster tries to access an external IP, the plugin
performs different Network Address Translation (NAT) manipulation on the traffic as it does not belong to the local cluster. Similarly, Submariner also programs certain iptables rules and it requires these rules to be applied
prior to the ones programmed by the network plugin.
Submariner tries to preserve the source IP of the Pods for cross-cluster communication for visibility,
ease of debugging, and security purposes.
```

File: src/content/deployment/calico/_index.en.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+So, Submariner requires its iptable rules have higher precedence over the iptable rules that are programmed
+by the local Cluster CNI. With Calico, it was seen that Calico is pretty aggressive in inserting its rules
+at the top and this was causing issues with cross-cluster communication.
+However, we have a solution. We can instruct Calico, via IPPools, not to perform NAT'ing on certain subnets.
+In our case, its the Pod and Service CIDRs of the remote Clusters. Once the IPPools are configured in the
+clusters, Calico will not perform SNAT for the configured CIDRs and allows Submariner to support cross-cluster
+connectivity.
```
Comment (nyechiel) [inline src/content/deployment/calico/_index.en.md:20]:
```suggestion
So, Submariner requires its iptables rules to take precedence over the rules that are programmed
by the local cluster network plugin. On clusters deployed with Calico as the network plugin, it was seen that rules inserted by Calico take precedence over Submariner, causing issues with cross-cluster communication.
To make Calico compatible with Submariner, it needs to be configured, via IPPools, not to perform NAT on the subnets associated with the Pod and Service CIDRs of the remote clusters. Once the IPPools are configured in the
clusters, Calico will not perform NAT for the configured CIDRs and allows Submariner to support cross-cluster
connectivity.
```

File: src/content/deployment/calico/_index.en.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+{{% notice note %}}
+When using [Submariner Globalnet](../../architecture/globalnet) with Calico, please avoid the default
+globalnet-cidr (i.e., 169.254.0.0/16) as its used internally within Calico. You can explicitly specify
+a non-overlapping globalnet-cidr while deploying Submariner.
+{{% /notice %}}
+
+**Let's take an example**:
+If you have two clusters, East and West, with Calico CNI connected via Submariner, you can create the following IPPools.
```
Comment (nyechiel) [inline src/content/deployment/calico/_index.en.md:29]:
```suggestion
If you have two clusters, East and West, deployed with the Calico network plugin and connected via Submariner, you should create the following IPPools:
```

**Review by nyechiel — COMMENTED** [review]
Some style suggestions

File: src/content/deployment/calico/_index.en.md (hunk 6 above)
Comment (sridhargaddam) [reply src/content/deployment/calico/_index.en.md:12]: Done

File: src/content/deployment/calico/_index.en.md (hunk 7 above)
Comment (sridhargaddam) [reply src/content/deployment/calico/_index.en.md:20]: Done

**Review by sridhargaddam — COMMENTED** [review]
Thanks for the review comments @nyechiel

**Review by nyechiel — APPROVED** [review]
LGTM, @tpantelis can you take a pass on it please?

File: src/content/quickstart/_index.en.md [hunk 9]
```diff
@@ -72,10 +72,11 @@ this document.
 
 Presently, Submariner has been tested with the following CNI Plugins that leverage kube-proxy with iptables mode:
 
-* [openshift-sdn](https://github.com/openshift/sdn)
+* [Openshift-SDN](https://github.com/openshift/sdn)
```
Comment (nyechiel) [inline src/content/quickstart/_index.en.md:75]:
```suggestion
* [OpenShift-SDN](https://github.com/openshift/sdn)
```

File: src/content/deployment/calico/_index.en.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+networking within the cluster. When a Pod in the cluster tries to access an external IP, the plugin
+performs different Network Address Translation (NAT) manipulation on the traffic as it does not
+belong to the local cluster. Similarly, Submariner also programs certain iptables rules and it
+requires these rules to be applied prior to the ones programmed by the network plugin.
+Submariner tries to preserve the source IP of the Pods for cross-cluster communication for visibility,
+ease of debugging, and security purposes.
+
+So, Submariner requires its iptables rules to take precedence over the rules that are programmed
+by the local cluster network plugin. On clusters deployed with Calico as the network plugin, it was
+seen that rules inserted by Calico take precedence over Submariner, causing issues with cross-cluster
```
Comment (tpantelis) [inline src/content/deployment/calico/_index.en.md:16]:
```suggestion
On clusters deployed with Calico as the network plugin, the rules inserted by Calico take precedence over Submariner, causing issues with cross-cluster
```

File: src/content/deployment/calico/_index.en.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+{{% notice note %}}
+When using [Submariner Globalnet](../../architecture/globalnet) with Calico, please avoid the default
+globalnet-cidr (i.e., 169.254.0.0/16) as its used internally within Calico. You can explicitly specify
+a non-overlapping globalnet-cidr while deploying Submariner.
+{{% /notice %}}
+
+**Let's take an example**:
+If you have two clusters, East and West, deployed with the Calico network plugin and connected
+via Submariner, you should create the following IPPools:
```
Comment (tpantelis) [inline src/content/deployment/calico/_index.en.md:30]:
```suggestion
As an example, consider two clusters, East and West, deployed with the Calico network plugin and connected via Submariner. For cluster East, the Service CIDR is 100.93.0.0/16 and the Pod CIDR is 
10.243.0.0/16. For cluster West, they are 100.92.0.0/16 and 10.242.0.0/16. The following IPPools
should be created:
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: src/content/deployment/calico/_index.en.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
++++
+title =  "Calico CNI"
+weight = 30
++++
+
+Typically, the Kubernetes network plugin (based on kube-proxy) programs iptables rules for Pods
```
[the remark this answers is not in the document]
Comment (sridhargaddam) [reply src/content/deployment/calico/_index.en.md]: Done

File: src/content/deployment/calico/_index.en.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+networking within the cluster. When a Pod in the cluster tries to access an external IP, the plugin
```
[the remark this answers is not in the document]
Comment (sridhargaddam) [reply src/content/deployment/calico/_index.en.md]: Done

**Review by sridhargaddam — COMMENTED** [review]
Thanks @tpantelis for your review comments.

**Review by mangelajo — APPROVED** [review]

**Comment by mangelajo** [comment]
Merging, I see comments requested by tom have been addressed.

**Pull request merged**